src/tls/cert.h

```diff
@@ -2,8 +2,177 @@
 // Licensed under the Apache 2.0 License.
 #pragma once
 
-#ifdef TLS_PROVIDER_IS_MBEDTLS
-#  include "mbedtls/cert.h"
-#else
-#  include "openssl/cert.h"
-#endif
+#include "ca.h"
+#include "ds/logger.h"
+#include "tls/tls.h"
+
+#include <cstring>
+#include <memory>
+#include <openssl/ssl.h>
+#include <optional>
+
+using namespace crypto;
+
+namespace tls
+{
+  enum Auth
+  {
+    auth_default,
+    auth_none,
+    auth_optional,
+    auth_required
+  };
+
+  // This class represents the authentication/authorization context for a TLS
+  // session. At least, it contains the peer's CA. At most, it also contains our
+  // own private key/certificate which will be presented in the TLS handshake.
+  // The peer's certificate verification can be overridden with the auth
+  // parameter.
+  class Cert
+  {
+  private:
+    std::shared_ptr<CA> peer_ca;
+    std::optional<std::string> peer_hostname;
+
+    crypto::OpenSSL::Unique_X509 own_cert;
+    crypto::OpenSSL::Unique_PKEY own_pkey;
+    bool has_own_cert;
+
+    Auth auth;
+
+  public:
+    Cert(
+      std::shared_ptr<CA> peer_ca_,
+      const std::optional<crypto::Pem>& own_cert_ = std::nullopt,
+      const std::optional<crypto::Pem>& own_pkey_ = std::nullopt,
+      CBuffer pw = nullb,
+      Auth auth_ = auth_default,
+      const std::optional<std::string>& peer_hostname_ = std::nullopt) :
+      peer_ca(peer_ca_),
+      peer_hostname(peer_hostname_),
+      has_own_cert(false),
+      auth(auth_)
+    {
+      crypto::OpenSSL::Unique_X509 tmp_cert;
+      crypto::OpenSSL::Unique_PKEY tmp_pkey;
+
+      if (own_cert_.has_value() && own_pkey_.has_value())
+      {
+        BIO* certBio = BIO_new(BIO_s_mem());
+        BIO_write(certBio, own_cert_->data(), own_cert_->size());
+        X509* cert = PEM_read_bio_X509(certBio, NULL, NULL, NULL);
+        BIO_free(certBio);
+        if (!cert)
+        {
+          auto err_str = tls::error_string(ERR_get_error());
+          LOG_FAIL_FMT("Cert::ctor: Could not parse certificate: {}", err_str);
+          throw std::logic_error("Could not parse certificate: " + err_str);
+        }
+        tmp_cert.reset(cert);
+
+        BIO* pkBio = BIO_new(BIO_s_mem());
+        BIO_write(pkBio, own_pkey_->data(), own_pkey_->size());
+        EVP_PKEY* pk = PEM_read_bio_PrivateKey(pkBio, NULL, NULL, NULL);
+        BIO_free(pkBio);
+        if (!pk)
+        {
+          auto err_str = tls::error_string(ERR_get_error());
+          LOG_FAIL_FMT("Cert::ctor: Could not parse key: {}", err_str);
+          throw std::logic_error("Could not parse key: " + err_str);
+        }
+        tmp_pkey.reset(pk);
+
+        has_own_cert = true;
+      }
+
+      own_cert = std::move(tmp_cert);
+      own_pkey = std::move(tmp_pkey);
+
+      if (pw.n)
+      {
+        // We don't seem to be using PW anywhere in CCF, so we should
```

Comment:
TBD?

Reply:
Will we ever use password-protected certs? If not, I agree, we can kill this branch

Reply:
No, but I'm not sure how things would work if we did?

Reply:
Yeah, too much complication and the code is easy to add back if we ever decide to go down that path.